src/main/java/org/jenkinsci/plugins/workflow/cps/CpsFlowExecution.java

```diff
@@ -334,6 +380,40 @@ private Object readResolve() {
     }
 
     /**
+     * Record time taken during a certain class of operation in this build.
+     * @param kind what sort of operation is being done
+     * @param end false to start, true at the end (pair in a {@code finally}-block)
+     */
+    synchronized void time(@Nonnull TimingKind kind, boolean end) {
+        if (timings == null) {
+            timings = new HashMap<>();
+        }
+        if (starts == null) {
+            starts = new HashMap<>();
+        }
+        if (end) {
+            Long orig = timings.get(kind.name());
+            if (orig == null) {
+                orig = 0L;
+            }
+            assert starts.containsKey(kind) : kind + " not in " + starts;
+            timings.put(kind.name(), orig + System.nanoTime() - starts.get(kind));
+        } else {
+            starts.put(kind, System.nanoTime());
+        }
+    }
+
+    synchronized void logTimings() {
```

Comment:
If the log level is not right, calling `log` is a no-op, so that makes no sense.

I could switch to a more specific logger though, so you could turn just this on.